common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts

```diff
@@ -45,4 +48,56 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
 
         return null;
     }
+
+    private _getACRToken(AADToken: string, retryCount: number, timeToWait: number): Q.Promise<string> {
+        let deferred = Q.defer<string>();
+        let tenantID = tl.getEndpointAuthorizationParameter(this.endpointName, 'tenantid', true);
+        let webRequest = new webClient.WebRequest();
+        webRequest.method = "POST";
+        const retryLimit = 5
+        webRequest.uri = `https://${this.registryURL}/oauth2/exchange`;
+        webRequest.body = (
+            `grant_type=access_token&service=${this.registryURL}&tenant=${tenantID}&access_token=${AADToken}`
+        );
+        webRequest.headers = {
+            "Content-Type": "application/x-www-form-urlencoded"
+        };
+        webClient.sendRequest(webRequest).then(
+            (response: webClient.WebResponse) => {
+                if (response.statusCode === 200) {
+                    deferred.resolve(response.body.refresh_token);
+                }
+                else if (response.statusCode == 429 || response.statusCode == 500) {
+                    if (retryCount < retryLimit) {
+                        let waitedTime = 2000 + timeToWait * 2;
+                        retryCount += 1;
+                        setTimeout(() => {
+                            deferred.resolve(this._getACRToken(AADToken, retryCount, waitedTime));
+                        }, waitedTime);
+                    }
+                    else {
+                        deferred.reject(tl.loc('CouldNotFetchAccessTokenforACRStatusCode', response.statusCode, response.statusMessage));
+                    }
+
+                }
+                else {
+                    deferred.reject(tl.loc('CouldNotFetchAccessTokenforMSIDueToACRNotConfiguredProperlyStatusCode', response.statusCode, response.statusMessage));
+                }
+            },
+            (error) => {
+                deferred.reject(error)
+            }
+        );
+        return deferred.promise;
+    }
+
+    public async getMSIAuthenticationToken(retryCount: number, timeToWait: number): Promise<RegistryAuthenticationToken> {
```

Comment:
This is kind of OCP and SRP violation. Have you considered creating a new TokenProvider for MSI and have a factory which returns SPN or MSI Token Provider based on Service connection Auth type ? 

I would like to evaluate if that requires significant rework here

Reply:
I don't think that would require significant rework. I could use the getToken() method inside of arm-common-v2 and use that as a baseline. Small amount of rework, but it could work better in our case.